pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,253 @@
+	}()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		equal = true
+		return nil
+	} else if ok {
+		action = resources.MetricUpdate
+	} else {
+		action = resources.MetricCreate
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroup.Name, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, action, err == nil)
+
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
+	cidrGroupName string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	delete(cidrGroupCache, cidrGroupName)
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, resources.MetricDelete, err == nil)
+	k.K8sEventReceived(apiGroup, metricLabel, resources.MetricDelete, true, true)
+
+	return err
+}
+
+func (k *K8sWatcher) updateCIDRGroupRefPolicies(
+	cidrGroup string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	var errs []error
+	for key, cnp := range cnpCache {
+		if !hasCIDRGroupRef(cnp, cidrGroup) {
+			continue
+		}
+
+		log.WithFields(logrus.Fields{
+			logfields.CiliumNetworkPolicyName: cnp.Name,
+			logfields.K8sAPIVersion:           cnp.APIVersion,
+			logfields.K8sNamespace:            cnp.Namespace,
+			logfields.CIDRGroupRef:            cidrGroup,
+		}).Info("Referenced CiliumCIDRGroup updated or deleted, recalculating CiliumNetworkPolicy rules")
+
+		initialRecvTime := time.Now()
+
+		// We need to deepcopy this structure because we are writing
+		// fields.
+		// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
+		cnpCpy := cnp.DeepCopy()
+
+		translationStart := time.Now()
+		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
+		metrics.CIDRGroupTranslationTimeStats.Observe(time.Since(translationStart).Seconds())
+
+		err := k.updateCiliumNetworkPolicyV2(cs, cnpCpy, translatedCNP, initialRecvTime)
+		if err == nil {
+			cnpCache[key] = cnpCpy
+		}
+
+		errs = append(errs, err)
+	}
+	return errors.Join(errs...)
+}
+
+func resolveCIDRGroupRef(cnp *types.SlimCNP, cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup) *types.SlimCNP {
+	refs := getCIDRGroupRefs(cnp)
+	if len(refs) == 0 {
+		return cnp
+	}
+
+	cidrsSets, err := cidrGroupRefsToCIDRsSets(refs, cidrGroupCache)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+			logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+			logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+			logfields.CIDRGroupRefs:           refs,
+		}).WithError(err).Warning("unable to translate all cidr groups to cidrs")
```

Comment:
nit, I think that errors are supposed to be all lower case since they're concatenated with other strings, but messages should start with a capital as though it's a sentence (this is at least how it seems like most messages are written in the codebase so far, from a quick glance at grep output):

```suggestion
		}).WithError(err).Warning("Unable to translate all cidr groups to cidrs")
```